hal/src/stm32f2xx/storage_hal.cpp

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2021 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "storage_hal.h"
+#include "flash_mal.h"
+#include "platform_config.h"
+#include "system_error.h"
+#include "spi_flash.h"
+#include "flash_mal.h"
+#include "check.h"
+#include <algorithm>
+
+int hal_storage_read(hal_storage_id id, uintptr_t addr, uint8_t* buf, size_t size) {
+    if (id == HAL_STORAGE_ID_INTERNAL_FLASH) {
+        memcpy(buf, (void*)addr, size);
+        return size;
+    }
+#ifdef USE_SERIAL_FLASH
+    else if (id == HAL_STORAGE_ID_EXTERNAL_FLASH) {
+        sFLASH_ReadBuffer(buf, addr, size);
+        return size;
+    }
+#endif // USE_SERIAL_FLASH
+
+    return SYSTEM_ERROR_NOT_FOUND;
+}
+
+int hal_storage_write(hal_storage_id id, uintptr_t addr, const uint8_t* buf, size_t size) {
+    if (id == HAL_STORAGE_ID_INTERNAL_FLASH) {
+        FLASH_Unlock();
+        FLASH_ClearFlags();
+        /* Data received are Word multiple */
+        FLASH_Status status = FLASH_COMPLETE;
+        for (size_t pos = 0; pos < size && status == FLASH_COMPLETE; pos += sizeof(uint32_t)) {
+            uint32_t tmp = 0xffffff;
```

Comment:
Yeah, you are right on the unaligned writes. As this was only for tests I was mainly focusing on whatever is required there in the context of this PR, but since you are running into issues in your tests, I'll quickly refactor to use existing functionality in `flash_common.cpp` for Gen 3 which tackles alignment.

Reply:
For visibility: the issue in my tests was caused by a bug in this code that uses `0xffffff` instead `0xffffffff` to initialize `tmp`, not by an unaligned write. Big thumbs up for still refactoring the code to use the functionality from flash_common.cpp :+1:

Reply:
Refactored to use `flash_common.cpp`